# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [1]:
# Your code here
import pandas as pd

In [2]:
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"

customer_data = pd.read_csv(url)
customer_data.head()

,Customer,ST,GENDER,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323


In [3]:
customer_data.shape

(4008, 11)

In [4]:
customer_data.dtypes

Customer                         str
ST                               str
GENDER                           str
Education                        str
Customer Lifetime Value          str
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints        str
Policy Type                      str
Vehicle Class                    str
Total Claim Amount           float64
dtype: object

In [5]:
customer_data.nunique()

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

In [6]:
print("ST:", customer_data["ST"].unique())
print()

print("GENDER:", customer_data["GENDER"].unique())
print()

print("Education:", customer_data["Education"].unique())
print()

print("Policy Type:", customer_data["Policy Type"].unique())
print()

print("Vehicle Class:", customer_data["Vehicle Class"].unique())
print()

ST: <StringArray>
['Washington',    'Arizona',     'Nevada', 'California',     'Oregon',
       'Cali',         'AZ',         'WA',          nan]
Length: 9, dtype: str

GENDER: <StringArray>
[nan, 'F', 'M', 'Femal', 'Male', 'female']
Length: 6, dtype: str

Education: <StringArray>
[              'Master',             'Bachelor', 'High School or Below',
              'College',            'Bachelors',               'Doctor',
                    nan]
Length: 7, dtype: str

Policy Type: <StringArray>
['Personal Auto', 'Corporate Auto', 'Special Auto', nan]
Length: 4, dtype: str

Vehicle Class: <StringArray>
['Four-Door Car',  'Two-Door Car',           'SUV',    'Luxury SUV',
    'Sports Car',    'Luxury Car',             nan]
Length: 7, dtype: str



In [11]:
customer_data["Customer Lifetime Value"].head(10)

0            NaN
1     697953.59%
2    1288743.17%
3     764586.18%
4     536307.65%
5     825629.78%
6     538089.86%
7     721610.03%
8    2412750.40%
9     738817.81%
Name: Customer Lifetime Value, dtype: str

In [10]:
customer_data["Number of Open Complaints"].head(10)

0    1/0/00
1    1/0/00
2    1/0/00
3    1/0/00
4    1/0/00
5    1/0/00
6    1/0/00
7    1/0/00
8    1/0/00
9    1/0/00
Name: Number of Open Complaints, dtype: str

In [9]:
customer_data["Number of Open Complaints"].unique()

<StringArray>
['1/0/00', '1/2/00', '1/1/00', '1/3/00', '1/5/00', '1/4/00', nan]
Length: 7, dtype: str

In [12]:
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"

customer_data = pd.read_csv(url)

In [13]:
customer_data["Number of Open Complaints"].unique()

<StringArray>
['1/0/00', '1/2/00', '1/1/00', '1/3/00', '1/5/00', '1/4/00', nan]
Length: 7, dtype: str

In [14]:
customer_data["Customer Lifetime Value"] = pd.to_numeric(
    customer_data["Customer Lifetime Value"].str.replace("%", "")
)

customer_data["Number of Open Complaints"] = customer_data["Number of Open Complaints"].str.extract(
    r"1/(\d+)/00"
).astype(float)

In [15]:
customer_data.dtypes

Customer                         str
ST                               str
GENDER                           str
Education                        str
Customer Lifetime Value      float64
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints    float64
Policy Type                      str
Vehicle Class                    str
Total Claim Amount           float64
dtype: object

In [16]:
print("Customer Lifetime Value:", customer_data["Customer Lifetime Value"].min(), "-", customer_data["Customer Lifetime Value"].max())
print()

print("Income:", customer_data["Income"].min(), "-", customer_data["Income"].max())
print()

print("Monthly Premium Auto:", customer_data["Monthly Premium Auto"].min(), "-", customer_data["Monthly Premium Auto"].max())
print()

print("Number of Open Complaints:", customer_data["Number of Open Complaints"].min(), "-", customer_data["Number of Open Complaints"].max())
print()

print("Total Claim Amount:", customer_data["Total Claim Amount"].min(), "-", customer_data["Total Claim Amount"].max())

Customer Lifetime Value: 200435.07 - 5816655.35

Income: 0.0 - 99960.0

Monthly Premium Auto: 61.0 - 35354.0

Number of Open Complaints: 0.0 - 5.0

Total Claim Amount: 0.382107 - 2893.239678


In [17]:
customer_data.describe().T

,count,mean,std,min,25%,50%,75%,max
Customer Lifetime Value,1068.0,793690.259345,643478.410278,200435.070000,403407.995000,588174.235000,896287.1975,5.816655e+06
Income,1071.0,39295.701214,30469.427060,0.000000,14072.000000,36234.000000,64631.0000,9.996000e+04
Monthly Premium Auto,1071.0,193.234360,1601.190369,61.000000,68.000000,83.000000,109.5000,3.535400e+04
Number of Open Complaints,1071.0,0.394024,0.883769,0.000000,0.000000,0.000000,0.0000,5.000000e+00
Total Claim Amount,1071.0,404.986909,293.027260,0.382107,202.157702,354.729129,532.8000,2.893240e+03


In [18]:
customer_data.mode(numeric_only=True)

,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Total Claim Amount
0,251459.20,0.0,65.0,0.0,321.6
1,445811.34,NaN,NaN,NaN,NaN


In [19]:
categorical_columns = ["ST", "GENDER", "Education", "Policy Type", "Vehicle Class"]

customer_data[categorical_columns].describe().T

,count,unique,top,freq
ST,1071,8,Oregon,320
GENDER,954,5,F,457
Education,1071,6,Bachelor,324
Policy Type,1071,3,Personal Auto,780
Vehicle Class,1071,6,Four-Door Car,576


### Conclusions

The dataset has 4,008 rows and 11 columns. Most of the columns already have suitable data types, but `Customer Lifetime Value` and `Number of Open Complaints` had to be converted from strings into numerical values so they could be properly analyzed.

I also noticed some inconsistencies in the categorical data. For example, the `ST` column uses both `California` and `Cali`, as well as `Arizona` and `AZ`. There are similar differences in the `GENDER` and `Education` columns, such as `F`, `Femal`, `female` and `Bachelor`, `Bachelors`. These values could be cleaned and standardized.

Looking at the numerical variables, there is quite a lot of variation in some of them. `Customer Lifetime Value` ranges from about 200,435 to 5.82 million, while `Income` ranges from 0 to almost 100,000. `Monthly Premium Auto` also has a very large range, from 61 to 35,354. In comparison, `Number of Open Complaints` only ranges from 0 to 5.

The mean and median also give us some useful information. For example, `Customer Lifetime Value` has a mean of about 793,690 but a median of about 588,174. The higher mean suggests that some very large values are affecting the average. The same can be seen with `Monthly Premium Auto`, where the mean is about 193 compared with a median of 83.

For the categorical variables, Oregon is the most common state, `F` is the most common gender, `Bachelor` is the most common education level, `Personal Auto` is the most common policy type, and `Four-Door Car` is the most common vehicle class.

Finally, there are some missing values in the dataset. `Customer Lifetime Value` has 3 missing values and `GENDER` has 117 missing values. These missing values, together with the inconsistent category names, would be useful to address during the data-cleaning stage before doing further analysis.

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [20]:
# Your code here
customer_locations = customer_data["ST"].value_counts()

customer_locations.sort_values().head(5)

ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [21]:
# Your code here
policy_counts = customer_data["Policy Type"].value_counts()

policy_counts

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

In [22]:
policy_counts.idxmax()

'Personal Auto'

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [25]:
# Your code here
Personal_Auto = customer_data.loc[customer_data["Policy Type"] == "Personal Auto"]

Corporate_Auto = customer_data.loc[customer_data["Policy Type"] == "Corporate Auto"]



In [26]:
Personal_Auto["Income"].mean()

np.float64(38180.69871794872)

In [30]:
Corporate_Auto["Income"].mean()


np.float64(41390.31196581197)

In [31]:
print("Average income for Personal Auto:", Personal_Auto["Income"].mean())
print()

print("Average income for Corporate Auto:", Corporate_Auto["Income"].mean())

Average income for Personal Auto: 38180.69871794872

Average income for Corporate Auto: 41390.31196581197


In [33]:
# i rounded up the figure to 2 decimal places

print("Personal Auto:", round(Personal_Auto["Income"].mean(), 2))
print("Corporate Auto:", round(Corporate_Auto["Income"].mean(), 2))

Personal Auto: 38180.7
Corporate Auto: 41390.31


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [34]:
# Your code here
customer_data["Total Claim Amount"].quantile(0.75)

np.float64(532.8)

In [39]:
high_claim_amount = customer_data[customer_data["Total Claim Amount"] > customer_data["Total Claim Amount"].quantile(0.75)]

high_claim_amount.shape



(264, 11)

In [38]:
high_claim_amount.describe().T

,count,mean,std,min,25%,50%,75%,max
Customer Lifetime Value,264.0,908686.813144,638767.789836,228759.69,472395.792500,778176.825000,1.054430e+06,3.844586e+06
Income,264.0,23677.344697,27013.483721,0.00,0.000000,18807.000000,4.242375e+04,9.931600e+04
Monthly Premium Auto,264.0,165.193182,623.930992,63.00,99.000000,114.000000,1.332500e+02,1.020200e+04
Number of Open Complaints,264.0,0.416667,0.943593,0.00,0.000000,0.000000,0.000000e+00,5.000000e+00
Total Claim Amount,264.0,782.228263,292.751640,537.60,606.521741,679.597985,8.514000e+02,2.893240e+03
